### ISSUE TITLE
Block theme issue with category template after changing a category name

### ISSUE THREAD

**Issue by hagege** [WordPress/gutenberg#56073]
### Description

If I have changed a category name in Posts > Categories, the corresponding template is defective because the category can no longer be found. In this case, I first have to adapt the template accordingly. 

I noticed the issue with Twenty Twenty-Four, but it should affect all block themes. 

It would be nice if, when the category name is changed, any template for the category is also adapted accordingly.

### Step-by-step reproduction instructions

![issue_renaming_category_name_and_custom_template](https://github.com/WordPress/gutenberg/assets/27292550/4fc53b6c-7c5b-41cd-ac42-8d1cf8067e62)

### Environment info

- WordPress 6.4.1
- Twenty Twenty-Four
- PHP 8.1
- Gutenberg not installed

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by franzaurus** [comment]
I found this to be an old and persistent issue. This bug happens not only to WP 6.4.1 but to 6.1.1, 6.2.2 and 6.3.2 as well. However, inspite of the bug the renamed category page always loads correctly.

All the testing environments don't have Gutenberg installed. For the previous releases test, I used the default TT3 theme, but using Artly theme for the WP 6.4.1 testing. This is to confirm that the issue affected all themes.

### On WP 6.1.1
![specific-category-template-bug-wp-611](https://github.com/WordPress/gutenberg/assets/79677758/7fdd7fc4-d54a-4176-81c9-a70b9d78e2fa)

### On WP 6.2.2
![specific-category-template-bug-wp-622](https://github.com/WordPress/gutenberg/assets/79677758/be1e8f94-96a5-4d42-984f-aaabb9cb99e3)

### On WP 6.3.2
![specific-category-template-bug-wp-632](https://github.com/WordPress/gutenberg/assets/79677758/382dfc04-4eae-4d6a-8657-21728ac6172d)

### On WP 6.4.1
![specific-category-template-bug-wp-641](https://github.com/WordPress/gutenberg/assets/79677758/c5da503b-dd39-4a93-b6c5-7c7fa5970a1e)

**Comment by justintadlock** [comment]
I'm torn about this. If I change a category/term name (templates are technically tied to the slug, not the name), it doesn't necessarily mean that I also want to the template to be matched to the changed category. But there are cases where it makes sense to do so.

I think a good balance would be a warning message in the Edit Category/Term screen and an option to also change the template slug.

Where the template exists is also a consideration. If the template is only in the theme (not a user template saved in the database), you can't change its slug. The process would actually need to copy the original template and give it a new slug based on the category/term.

**Comment by franzaurus** [comment]
> templates are technically tied to the slug, not the name

Oh, I see. I tried this by changing the category/term name but left its slug intact (Unknown category/term name with uncategorized slug). I noticed that the template section in the Site Editor no longer informed me that the template category 'Uncategorized' was not found. Instead, it properly listed 'Category: Unknown template' there. So yeah, it could be used as a workaround.

However, while I understand that a theme can include a category/term template, isn't a specific category/term template always user-created? In this case, I can't imagine why a user would want to change a category/term name without also changing the slug.

**Comment by hagege** [comment]
> However, while I understand that a theme can include a category/term template, isn't a specific category/term template always user-created? In this case, I can't imagine why a user would want to change a category/term name without also changing the slug.

In fact, I see it the same way and would have assumed in any case that the template would also be adapted accordingly.

**Comment by bph** [comment]
The handling of category changes is inconsistent, and results in surprising behavior. I observed the following: 

**Change the name of the category only**, with the slug untouched, the template name displays the new name througout the editor and on the front page. 

https://github.com/WordPress/gutenberg/assets/39980/43753782-8b38-44fc-8903-b36074b8c9c2


**Change the name of the category and the slug,** the template name  doesn't recognize the category and displays "Category not found" message. Once the slug is changed back to the previous version, the template is found and works. 

https://github.com/WordPress/gutenberg/assets/39980/229da5d1-a11d-4804-a0eb-b585b21a3075

My suggestion would be to solve it the WordPress way and connect the template with the categoriy via the categoryID instead of the slug to show consistent behavior when values change.

**Comment by justintadlock** [comment]
> However, while I understand that a theme can include a category/term template, isn't a specific category/term template always user-created? 

Category templates were always created in the theme throughout most of WordPress' history. The ability for a user to create them from the UI is a very recent feature.

> In this case, I can't imagine why a user would want to change a category/term name without also changing the slug.

The primary reasons would be not breaking existing links (the category slug is used in the URL) and SEO (keep the same URL but update the category page title).

> My suggestion would be to solve it the WordPress way and connect the template with the categoriy via the categoryID instead of the slug to show consistent behavior when values change.

That sounds like a good approach. Both the core category and tag taxonomies already support ID-based templates. Other taxonomy term archives do not support them, so core would need to add `taxonomy-{term_taxonomy}-{term_id}.html` in the [taxonomy term hierarchy](https://developer.wordpress.org/themes/templates/template-hierarchy/#taxonomy-term-hierarchy).

There'd also need to be a mechanism to handle back-compat with currently-saved templates based on the slug. 

I'm still not sure how I'd approach this if the template is only in the theme (not created by the user). Just leave everything as is? Maybe a warning message when changing the slug (honestly, I think we should have that anyway)?

And---not to expand this issue too much---this approach would make user-created term templates inconsistent with single post templates. We should consider using the ID for them too (only pages currently support ID-based templates at the template level).

The biggest potential hurdle I see is with exporting the theme. If someone is building the theme on one install and plans to export it and import into another install, the IDs for the terms are going to be different. It's primarily for this reason that I'd like to solve this while sticking to slugs. Otherwise, we could potentially be adding additional levels of complexity.

**Comment by nullbytes** [comment]
@justintadlock so, coming back to your suggestion:
> Maybe a warning message when changing the slug (honestly, I think we should have that anyway)?
Can we add this warning message to move this issue forward?

**Comment by jameskoster** [comment]
A couple of ideas to indirectly allow users an easy way to address this without touching the template naming logic;

1. Allow theme-supplied term-specific templates to be duplicated. On duplication the user would be presented with the modal where they select which term the template should be applied to.
2. Allo user-generated term-specific templates to be 'remapped' (https://github.com/WordPress/gutenberg/issues/48645) which uses the same flow.

**Comment by justintadlock** [comment]
@WordPress/outreach - Just looking for feedback on how you think the UX should work in this scenario.

**Comment by markhowellsmead** [comment]
I agree that a warning should be displayed when editing the slug in the taxonomy editor. If the theme doesn't contain a template for that specific slug, then no warning is needed. If there is, then this should be renamed only if the user agrees to it.

As long as the warning message with yes/no option is clear and understandable for non-technical users, then that would be the best way to go. It should be clear to the user that selecting "no" may (will) have an effect on how the taxonomy list view is displayed in the frontend. If they then choose to leave the template slug unchanged, it's been because of an informed decision.

**Comment by bacoords** [comment]
A few thoughts on this:

- I wouldn't personally be in favor of tying templates to category terms via an ID unless that's actually the way it's shown in the UI. The UI should as closely match the path you're taking down the template heirarchy as possible. Theme developers can rely on slugs being persistent but not IDs, so that approach should be used sparingly.
- A warning when changing a slug would be great - and would be great across all of the site editor and post types. Changing a slug also changes the URL of an object, which is something every user should be aware of (and some SEO plugins I believe do add a warning for). Something like "Warning, changing the slug will affect the URL of this category as well as disconnecting it from [] template and reverting it to the [default] template."
